File: leetcode/18_Sum.java
// URL: https://leetcode.com/problems/4sum/
//
// Problem: 18. 4Sum
//

// Solution:
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> result = new ArrayList<>();
        Set<List<Integer>> set = new HashSet<>();
        Arrays.sort(nums);
        for(int i = 0 ; i < nums.length ; i++){
            for(int j = i + 1; j < nums.length ; j++){
                int k = j + 1;
                int l = nums.length - 1;
                while (k < l){
                    long total = (long) nums[i] + (long) nums[j] + (long) nums[k] + (long) nums[l];
                    if (total == target){
                        List<Integer> pair = Arrays.asList(nums[i],nums[j],nums[k],nums[l]);
                        set.add(pair);
                        k++;
                        l--;
                    }
                    else if(total < target){
                        k++;
                    }
                    else{
                        l--;
                    }
                }
            }
        }
        result.addAll(set);
        return result;
    }
}

File: leetcode/18_Sum.py
# URL: https://leetcode.com/problems/4sum/
#
# Problem: 18. 4Sum
#

# Solution:
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        s = set()
        num = []
        for i in range (len(nums)-3):
            for j in range (i+1,len(nums)):
                k = j + 1
                l = len(nums) - 1
                while k < l:
                    sums = nums[i] + nums[j] + nums[k] + nums[l]
                    if sums == target:
                        s.add(tuple(sorted((nums[i],nums[j],nums[k],nums[l]))))
                        k += 1
                        l -= 1
                    elif sums < target:
                        k += 1
                    else:
                        l -= 1
        for i in s:
            num.append(list(i))
        return num